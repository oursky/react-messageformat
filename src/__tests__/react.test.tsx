import React, { useContext } from "react";
import { create } from "react-test-renderer";
import {
  LocaleProvider,
  Context,
  FormattedMessage,
  MessageOwnProps,
} from "../react";

const locale = "en";
const messageByID = {
  "plain.string": "Hello world",
  argument: "I meet {GUEST}",
  plural: "{N, plural, =4{four} one{ichi} other{#}}",
  "react.element": "{ANY} {NESTED} {REACT} {ELEMENT}",
  react: "This is a {A, react, href{{SCHEME}://{HOST}} children{link}}",
  "react.nonstring": "This is a {A, react, href{{href}}}",
  "react.imperative.string": "This is a {B, react}",
  "react.imperative.nested":
    "This is a {strong, react, children{{i, react, children{content}}}}",
  "regression.0":
    "Are you sure to change the {a, select, varianta{A} variantb{B} other{}}?",
};

function Shortcut(props: MessageOwnProps) {
  return (
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <FormattedMessage {...props} />
    </LocaleProvider>
  );
}

function Input(props: MessageOwnProps) {
  const { id, values } = props;
  const { renderToString } = useContext(Context);
  return <input placeholder={renderToString(id, values)} />;
}

class AnyComponent extends React.Component {
  render() {
    return "AnyComponent";
  }
}

test("plain.string", () => {
  const tree = create(<Shortcut id="plain.string" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("argument", () => {
  const tree = create(
    <Shortcut
      id="argument"
      values={{
        GUEST: "John",
      }}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("plural", () => {
  const tree = create(
    <Shortcut
      id="plural"
      values={{
        N: 4,
      }}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("React Element as value", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <FormattedMessage
        id="react.element"
        values={{
          ANY: <FormattedMessage id="plain.string" />,
          NESTED: <AnyComponent />,
          REACT: (
            <FormattedMessage
              id="argument"
              values={{
                GUEST: "John",
              }}
            />
          ),
          ELEMENT: (
            <FormattedMessage
              id="plural"
              values={{
                N: 1,
              }}
            />
          ),
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Embedded React Element", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <FormattedMessage
        id="react"
        components={{
          A: "a",
        }}
        values={{
          SCHEME: "https",
          HOST: "www.example.com",
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Embedded React Element with non-string value", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <FormattedMessage
        id="react.nonstring"
        components={{
          A: "a",
        }}
        values={{
          href: {
            valueOf: () => "https://reactjs.org",
            toString: () => "https://reactjs.org",
          },
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("imperative", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <Input id="plain.string" />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("imperative nested", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <Input id="react.imperative.nested" />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("defaultComponents", () => {
  const tree = create(
    <LocaleProvider
      locale={locale}
      messageByID={messageByID}
      defaultComponents={{
        A: "a",
      }}
    >
      <FormattedMessage
        id="react"
        values={{
          SCHEME: "https",
          HOST: "www.example.com",
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("components override defaultComponents", () => {
  const tree = create(
    <LocaleProvider
      locale={locale}
      messageByID={messageByID}
      defaultComponents={{
        A: "a",
      }}
    >
      <FormattedMessage
        id="react"
        components={{
          A: "span",
        }}
        values={{
          SCHEME: "https",
          HOST: "www.example.com",
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("regression.0", () => {
  const tree = create(
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <Input
        id="regression.0"
        values={{
          a: "variantb",
        }}
      />
    </LocaleProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
