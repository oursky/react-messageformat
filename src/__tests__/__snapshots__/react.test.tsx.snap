// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embedded React Element 1`] = `
Array [
  "This is a ",
  <a
    href="https://www.example.com"
  >
    link
  </a>,
]
`;

exports[`Embedded React Element with non-string value 1`] = `
Array [
  "This is a ",
  <a
    href={
      Object {
        "toString": [Function],
        "valueOf": [Function],
      }
    }
  />,
]
`;

exports[`React Element as value 1`] = `
Array [
  "Hello world",
  " ",
  "AnyComponent",
  " ",
  "I meet ",
  "John",
  " ",
  "ichi",
]
`;

exports[`argument 1`] = `
Array [
  "I meet ",
  "John",
]
`;

exports[`components override defaultComponents 1`] = `
Array [
  "This is a ",
  <span
    href="https://www.example.com"
  >
    link
  </span>,
]
`;

exports[`defaultComponents 1`] = `
Array [
  "This is a ",
  <a
    href="https://www.example.com"
  >
    link
  </a>,
]
`;

exports[`imperative 1`] = `
<input
  placeholder="Hello world"
/>
`;

exports[`imperative nested 1`] = `
<input
  placeholder="This is nested"
/>
`;

exports[`plain.string 1`] = `"Hello world"`;

exports[`plural 1`] = `"four"`;

exports[`regression.0 1`] = `
<input
  placeholder="Are you sure to change the B?"
/>
`;

exports[`regression.1 1`] = `
<input
  placeholder="Message shouldn't be HTML-escaped."
/>
`;
